sdk/storage/Azure.Storage.Common.DataMovement/src/FilesystemScanner.cs

```diff
@@ -5,108 +5,142 @@
 using System.Collections.Generic;
 using System.IO;
 using System.Text;
+using System.Linq;
 
 namespace Azure.Storage.Common.DataMovement
 {
     /// <summary>
     /// FilesystemScanner class.
     /// </summary>
-    public static class FilesystemScanner
+    public class FilesystemScanner
     {
         /// <summary>
-        /// Enumerates all files pointed to by the provided path, including those in subdirectories (if path is a directory).
+        /// Fully qualified path(s) equivalent to user-supplied path(s).
         /// </summary>
-        /// <param name="path">Filesystem location.</param>
-        /// <returns>Enumerable list of absolute paths containing all relevant files the user has permission to access.</returns>
-        public static IEnumerable<string> ScanLocation(string path)
+        public IEnumerable<string> FullPaths
         {
-            // Path type is ambiguous at start
-            bool isDirectory = false;
+            get;
+            private set;
+        }
 
+        /// <summary>
+        /// Single path constructor for FilesystemScanner.
+        /// </summary>
+        /// <param name="path"></param>
+        public FilesystemScanner(string path)
+        {
             try
             {
-                // Make sure we're dealing with absolute, well-formatted path
-                path = Path.GetFullPath(path);
+                // Ensure we're dealing with an absolute, well-formatted path
+                FullPaths = new string[] { Path.GetFullPath(path) };
+            }
+            catch
+            {
+                // TODO: Logging for bad path passed to constructor
+                throw;
+            }
+        }
 
-                // Check if path points to a directory
-                if ((File.GetAttributes(path) & FileAttributes.Directory) == FileAttributes.Directory)
-                {
-                    isDirectory = true;
-                }
+        /// <summary>
+        /// Multi-path constructor for FilesystemScanner.
+        /// </summary>
+        /// <param name="paths"></param>
+        public FilesystemScanner(IEnumerable<string> paths)
```

Comment:
What's the scenario somebody would provide collection of paths? 
I can imagine somebody could create scanner per path and aggregate results themselves.

Reply:
The work item for the scanner mentioned maybe allowing one scanner to traverse multiple paths, so I added that here.

If users will be making jobs and adding them to a transfer manager, and scanning is handled by those jobs, then maybe we could have this functionality of using multiple paths be handled by a job instead? In which case, said job would create multiple scanners. Or like you said, users could just make several individual jobs.

In either case, the scanner won't need to worry about this scenario (and it'll probably be easier to build from if not), so I'll let it just worry about a single path.

Reply:
I'd rather build separate component to farm out each directory to different scanner. Or maybe we'll build MultiPathScanner or so. 
Ability to scan multiple paths should be added when it's needed - this is when it becomes clear how to do this.

Reply:
@amnguye do we know why would we want to pack multi-path scanning into this thing ?

Reply:
Yeah I did put that in the work item. I probably didn't think further when I wrote that because taking in multiple paths would assume the scanner would be stateful.

What I want to attempt to do is have the scanner work on the current path given when it is given a path to scan, and if the user decides to add on more jobs (more paths to scan) then we continue to can the other paths from the queue. However I wanted to finish up adding in the job scheduler / transfer manager so that way we have an API to be able to funnel in these paths to use the scanner.

I'm comfortable getting rid of this API but eventually we will probably add an advanced version of this where it continues to scan, if there's paths to scan.

Reply:
So if we're doing multiple upload jobs with one transfer manager, we want all those jobs to use one scanner instead of one scanner per job? Then, are we waiting for all paths to be scanned and then hitting the file scheduler with all files, or taking a path, sending files to scheduler, taking a path, sending files to scheduler, so on? I'm confused on how we're separating the files of each job, and whether that's even a concern.

Would it be better suited to change functionality from storing paths in instance variables and scanning them on call, to instead keeping the found files in an instance variable and adding to it by adding paths?

Reply:
I think it will depend on how many threads we dedicate to the scanner portion of the process, which will determine how many scanners we have. So each job wont use the same scanner, unless we have more than one job. We are not waiting for all paths to be scanned before hitting the file scheduler, but we can implement that part in more detail once we get the file scheduler rolling.

Yes, eventually we should change it so that we are pushing these paths to the file scheduler so that we can start the upload and/or download operation.

Reply:
Let's make Scanner take one path in ctor for now and make it simple. We'll revisit that choice later when we build upper layers.